### PR TITLE
feat: add Message and Discoverer

### PR DESCRIPTION
**Pull request by gu18168**

### CODE REVIEW

File: internal/discoverer/discoverer.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+type Node struct {
+	host   string
+	weight int
+}
+
+type Service struct {
+	name     string
+	nodes    []Node
+	entities []string
```
Comment (spacewander) [inline internal/discoverer/discoverer.go]: Better to add a comment about the difference between nodes & entities?

File: internal/discoverer/discoverer.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	args     map[string]string
+}
+
+// queryDecode check and extract values from the query message
+func queryDecode(msg utils.Message) ([]string, map[string]string, error) {
+	msgLength := len(msg)
+	if msgLength < 3 {
```
Comment (spacewander) [inline internal/discoverer/discoverer.go]: Please avoid repeating the magic number 3

File: internal/discoverer/discoverer.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
+package discoverer
+
+import (
+	"errors"
+	"fmt"
+	"github.com/api7/apisix-seed/internal/utils"
```
Comment (spacewander) [inline internal/discoverer/discoverer.go]: Please configure your editor to solve the imports.

Comment (gu18168) [reply internal/discoverer/discoverer.go]: Done.

File: internal/discoverer/discoverer.go (hunk 1 above)
Comment (gu18168) [reply internal/discoverer/discoverer.go]: Done.

File: internal/discoverer/discoverer.go (hunk 2 above)
Comment (gu18168) [reply internal/discoverer/discoverer.go]: I have reorganized the query and watch messages about the communication between components.

File: internal/discoverer/discoverer.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	msg := make(utils.Message, 0, len(s.entities))
+	for _, entity := range s.entities {
+		msg.Add("entity", entity)
+	}
+	return msg
+}
+
+// EncodeWatch encodes a service to the watch message
+func (s *Service) EncodeWatch() (*comm.Watch, error) {
+	header, err := comm.NewWatchHeader([]string{utils.EventUpdate, s.name})
```
Comment (spacewander) [inline internal/discoverer/discoverer.go:56]: Is `utils.EventUpdate` here a placeholder?

Comment (gu18168) [reply internal/discoverer/discoverer.go:56]: In the current design, the watch messages sent by the discoverer are all listening for an update to a service, so for now the message event type is simply fixed to update.

**Review by spacewander — APPROVED** [review]

**Pull request merged**